### ISSUE TITLE
question on dynamically rendering an image

### ISSUE THREAD

**Issue by iamakimmer** [susanBuck/e28-fall20#24]
If I rely on some outside api to get an image after the page has loaded, the image will return a broken image first:
```
    <img id="logo"
      :src="somepicturethatillgetlater"
      alt="logo"
    />
```

Is it common practice to attach dynamic image or href attributes it to a v-if?
```
    <img id="logo"
      v-if="somepicturethatillgetlater"
      :src="somepicturethatillgetlater"
      alt="logo"
    />
```

**Comment by LDubya** [comment]
What are you showing before the image url has been generated? Maybe you'd want to have a loading image to show by default?
```js
{
   somepicturethatillgetlater : "img/defaultimage"
}
```
Then change the value from the default image url to the one provided by the outside API later?

Or if no image should be shown until one is provided by the api, 
I'd probably do the `<img v-if="somepicturethatillgetlater" />`

**Comment by iamakimmer** [comment]
Ohhh thats even better. Thanks!!

**Comment by kalbfled** [comment]
`<img v-if="somepicturethatillgetlater" />`

I think this will shift the content once the image loads, which probably isn't desirable.  Maybe use a placeholder div element that contains the image.

```css
div.missing
{
    heigh: <image height>px;
    width: <image width>px;
}
```

```html
<div class="missing">
    <img v-if="somepicturethatillgetlater" />
</div
```

**Comment by susanBuck** [comment]
@kalbfled - Good point about the shifting. Instead of surrounding the image in a div, though, you could just apply the height/width style to the image itself.

**Comment by LDubya** [comment]
Hmm 🤔, I'm thinking maybe @kalbfled is referring to when the `:src` is empty by default, where in combination with `v-if` the `< img />` element isn't yet rendered on the page? 

Is there something like `v-show`, but instead of toggling the `display` property it toggles the `visibility` property?

**Comment by kalbfled** [comment]
The goal of my markup is to avoid shifting and a broken image graphic when the source isn't available.

**Comment by susanBuck** [comment]
@LDubya - Ah yes, but I was thinking about your suggestion where he used a dummy/placeholder image - so something is rendered. : ) 

@kalbfled - Have you got this working how you're expecting, or still looking for more guidance/suggestions? If so, do you have an example we can see somewhere?

**Comment by kalbfled** [comment]
https://jsfiddle.net/x2pavkc7/

Note also that I corrected my previous markup post because `height` and `width` are not `div` attributes.

**Comment by susanBuck** [comment]
@kalbfled -

Unfortunately for your example, it looks like the image that's supposed to be blocked from hot-linking `https://blogs-images.forbes.com/robertglatter/files/2012/11/Smiley_Face61.png` is actually loading. : )

When I swapped in a bad image path in your code (e.g. `<img src='bad-image.png'/>`), the result was the space was still reserved, but you do see the broken image icon.

![image](https://user-images.githubusercontent.com/237711/96148253-14d18e00-0ed6-11eb-9b8b-5d83c0b55d03.png)

That got me looking for some other solutions and I came up with this: https://jsfiddle.net/susanbuck/vd3rw764/10/

It uses a Vue event listener to listen for an `error` event on the image, and if it occurs, it loads in a placeholder image. 

This example is building on the example you provided, where we're dealing with handling a broken image. But I'm wondering if that's actually what you're trying to address, as your original post said:

> If I rely on some outside api to get an image after the page has loaded, the image will return a broken image first:

If this is the case you're trying to address, I'd set the an imageSource data property on initial page load that points to a placeholder image, and then once the call to the outside API completes, imageSource gets updated.
